addons/membership/models/partner.py

```diff
@@ -0,0 +1,163 @@
+    membership_cancel = fields.Date(compute='_compute_membership_date', store=True, help='Date on which membership has been cancelled')
+
+    @api.depends('free_member', 'member_lines')
+    def _compute_membership_date(self):
+        """Return  date of membership"""
+        if self.ids:
+            ids_to_search = []
+            for partner in self:
+                if partner.associate_member:
+                    ids_to_search.append(partner.associate_member.id)
+                else:
+                    ids_to_search.append(partner.id)
+            self.env.cr.execute("""
+                SELECT
+                    p.id as id,
+                    MIN(m.date_from) as membership_start,
+                    MAX(m.date_to) as membership_stop,
+                    MIN(CASE WHEN m.date_cancel is not null THEN 1 END) as membership_cancel
+                FROM
+                    res_partner p
+                LEFT JOIN
+                    membership_membership_line m
+                    ON (m.partner = p.id)
+                WHERE
+                    p.id IN %s
+                GROUP BY
+                    p.id""", (tuple(ids_to_search), ))
+            for record in self.env.cr.dictfetchall():
+                partner = self.browse(record.pop('id')).update(record)
+
+    @api.depends('member_lines.account_invoice_id.state', 'membership_state', 'associate_member', 'free_member')
+    def _compute_membership_state(self):
+        return self._membership_state()
+
+    @api.constrains('associate_member')
+    def _check_associate_member(self):
+        """Check  Recursive  for Associated Members.
+        """
+        if not self._check_recursion(parent='associate_member'):
+            raise UserError(_('Error ! You cannot create recursive associated members.'))
+
+    @api.model
+    def _cron_update_membership(self):
+        partners = self.search([('membership_state', '=', 'paid')])
+        if partners:
+            partners._store_set_values(['membership_state'])
+
+    def _membership_state(self):
+        """This Function return Membership State For Given Partner. """
+        today_date = date.today()
+        for partner in self:
+            partner.membership_state = 'none'
+            if partner.membership_cancel and today_date > fields.Date.from_string(partner.membership_cancel):
+                partner.membership_state = 'free' if partner.free_member else 'canceled'
+                continue
+            if partner.membership_stop and today_date > fields.Date.from_string(partner.membership_stop):
+                partner.membership_state = 'free' if partner.free_member else 'old'
+                continue
+
+            state = 4
+            if partner.member_lines:
+                for mline in partner.member_lines:
+                    if fields.Date.from_string(mline.date_to) >= today_date:
+                        if mline.account_invoice_line and mline.account_invoice_line.invoice_id:
```

Comment:
if mline.account_invoice_line and mline.account_invoice_line.invoice_id: ==>  if mline.account_invoice_line.invoice_id: